File: backend/app/schemas/analytics.py
from pydantic import BaseModel
from typing import Optional, List
from uuid import UUID
from decimal import Decimal
from datetime import date


# ==================== CATEGORY ANALYTICS ====================
class CategorySpending(BaseModel):
    category: str
    amount: Decimal
    percentage: Decimal
    expense_count: int


# ==================== TIME SERIES ANALYTICS ====================
class SpendingDataPoint(BaseModel):
    date: date
    amount: Decimal
    expense_count: int


# ==================== MEMBER ANALYTICS ====================
class MemberContribution(BaseModel):
    user_id: UUID
    user_name: str
    profile_picture: Optional[str] = None
    total_paid: Decimal       # Total amount they paid for group
    total_share: Decimal      # Their share of expenses
    net_contribution: Decimal # paid - share (positive = overpaid)


# ==================== FRIEND ANALYTICS ====================
class FriendSpending(BaseModel):
    friend_id: UUID
    friend_name: str
    profile_picture: Optional[str] = None
    total_spent: Decimal
    expense_count: int


# ==================== GROUP ANALYTICS RESPONSE ====================
class GroupAnalyticsResponse(BaseModel):
    group_id: UUID
    group_name: str
    period: str
    start_date: date
    end_date: date
    total_spending: Decimal
    expense_count: int
    category_breakdown: List[CategorySpending]
    spending_over_time: List[SpendingDataPoint]
    member_contributions: List[MemberContribution]
    average_expense: Decimal
    top_category: Optional[str] = None


# ==================== FRIENDS ANALYTICS RESPONSE ====================
class FriendsAnalyticsResponse(BaseModel):
    period: str
    start_date: date
    end_date: date
    total_spending: Decimal
    expense_count: int
    category_breakdown: List[CategorySpending]
    spending_over_time: List[SpendingDataPoint]
    friend_breakdown: List[FriendSpending]
    average_expense: Decimal
    top_category: Optional[str] = None

File: frontend/src/types/analytics.ts
// Analytics Types

export type TimePeriod = '7d' | '30d' | '3m' | '1y';

export interface CategorySpending {
  category: string;
  amount: number;
  percentage: number;
  expense_count: number;
}

export interface SpendingDataPoint {
  date: string;
  amount: number;
  expense_count: number;
}

export interface MemberContribution {
  user_id: string;
  user_name: string;
  profile_picture?: string;
  total_paid: number;
  total_share: number;
  net_contribution: number;
}

export interface FriendSpending {
  friend_id: string;
  friend_name: string;
  profile_picture?: string;
  total_spent: number;
  expense_count: number;
}

export interface GroupAnalyticsResponse {
  group_id: string;
  group_name: string;
  period: string;
  start_date: string;
  end_date: string;
  total_spending: number;
  expense_count: number;
  category_breakdown: CategorySpending[];
  spending_over_time: SpendingDataPoint[];
  member_contributions: MemberContribution[];
  average_expense: number;
  top_category?: string;
}

export interface FriendsAnalyticsResponse {
  period: string;
  start_date: string;
  end_date: string;
  total_spending: number;
  expense_count: number;
  category_breakdown: CategorySpending[];
  spending_over_time: SpendingDataPoint[];
  friend_breakdown: FriendSpending[];
  average_expense: number;
  top_category?: string;
}
